test/Atmos/Parameterizations/Microphysics/single_eddy_test_Kessler.jl

```diff
@@ -0,0 +1,358 @@
+using CLIMA.Vtk
+
+using LinearAlgebra
+using StaticArrays
+using Printf
+
+using CLIMA.PlanetParameters
+using CLIMA.MoistThermodynamics
+using CLIMA.Microphysics
+
+@static if Base.find_package("CuArrays") !== nothing
+  using CUDAdrv
+  using CUDAnative
+  using CuArrays
+  const ArrayTypes = VERSION >= v"1.2-pre.25" ? (Array, CuArray) : (Array,)
+else
+  const ArrayTypes = (Array, )
+end
+
+const _nstate = 6
+const _ρ, _U, _W, _E, _qt, _qr = 1:_nstate
+const stateid = (ρid = _ρ, Uid = _U, Wid = _W, Eid = _E,
+                 qtid = _qt, qrid = _qr)
+const statenames = ("ρ", "U", "W", "E", "qt", "qr")
+
+const _nauxcstate = 2
+const _c_z = 1
+const _c_x = 2
+
+
+# preflux computation
+@inline function preflux(Q, _...)
+  DFloat = eltype(Q)
+
+  @inbounds ρ, U, W, qt, qr = Q[_ρ], Q[_U], Q[_W], Q[_qt], Q[_qr]
+
+  ρ_ground::DFloat = 1 #TODO ρ[0]
+
+  u, w = U / ρ, W / ρ
+  rain_w = terminal_velocity(qt, qr, ρ, ρ_ground)
+
+  # return 2 velocity components and rain fall speed for wave speed calculation
+  (u, w, rain_w)
+end
+
+
+# boundary condition
+@inline function bcstate!(QP, VFP, auxP, nM, QM, VFM, auxM, bctype, t,
+                          uM, wM, rain_wM)
+  @inbounds begin
+
+    UM, WM, EM = QM[_U],  QM[_W],  QM[_E]
+    qtM, qrM   = QM[_qt], QM[_qr]
+
+    UnM = nM[1] * UM + nM[2] * WM
+    QP[_U] = UM - 2 * nM[1] * UnM
+    QP[_W] = WM - 2 * nM[2] * UnM # TODO - what to do about rain fall speed?
+
+    QP[_E], QP[_qt], QP[_qr] = EM, qtM, qrM
+
+    auxM .= auxP
+
+    # To calculate uP and wP we use the preflux function
+    preflux(QP, auxP, t)
+
+    # Required return from this function is either nothing
+    # or preflux with plus state as arguments
+  end
+end
+
+
+# max eigenvalue
+@inline function wavespeed(n, Q, aux, t, u, w, rain_w)
+  @inbounds abs(n[1] * u + n[2] * max(w, rain_w, w+rain_w))
+end
+
+
+@inline function constant_auxiliary_init!(aux, x, z, _...)
+  @inbounds aux[_c_z] = z
+  @inbounds aux[_c_x] = x #TODO - tmp for printing
+end
+
+
+@inline function source!(S, Q, aux, t)
+  @inbounds begin
+    DFloat = eltype(Q)
+
+    ρ, E, U, W, qt, qr = Q[_ρ], Q[_E], Q[_U], Q[_W], Q[_qt], Q[_qr]
+    z = aux[_c_z]
+    x = aux[_c_x]
+
+    S .= 0
+    e_int = (E - 1//2 * (U^2 + W^2) - grav * z) / ρ
+    ts  =  PhaseEquil(e_int, qt, ρ)  # hidden saturation adjustment here
+    q_sat_adj = PhasePartition(ts)
+
+    timescale::eltype(Q) = 10
+
+    dqrdt = ql2qr(q_sat_adj.liq, timescale)
+
+    S[_qr]  = dqrdt
+    #S[_qt] -= dqrdt
+    #S[_E]  -= dqrdt * DFloat(e_int_v0) # TODO - move to microphysics sources
+
+    # TODO add rain evaporation
+
+    #if x == 0 && z >= 750
+    #  @printf("z = %4.2f qt = %.8e ql = %.8e qr = %.8e dqrdt = %.8e \n", z, qt, q_sat_adj.liq, qr, dqrdt)
+    #  if z == 1500
+    #      @printf("  ")
+    #  end
+    #end
+
+    #if qt < 0
+    #    @show(qt)
+    #end
+
+  end
+end
+
+
+# physical flux function
+eulerflux!(F, Q, QV, aux, t) = eulerflux!(F, Q, QV, aux, t, preflux(Q)...)
```

Comment:
I haven't had a chance to look deeply into this, but the error you are seeing (if the error location is correct), means that either `eulerflux!` or `preflux` is causing an issue.
Maybe try removing the `_....` from preflux?